pkg/offline/offline.go

```diff
@@ -77,40 +78,81 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 	}, nil
 }
 
-// Sync returns a function to send queued heartbeats to the WakaTime API.
+// Sync returns a function to send all queued heartbeats to the WakaTime API.
 func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
 	return func(next heartbeat.Handle) error {
 		log.Debugf("execute offline sync with file %s", filepath)
 
-		hh, err := popHeartbeats(filepath, syncLimit)
-		if err != nil {
-			return fmt.Errorf("failed to fetch heartbeat from offline queue: %s", err)
-		}
+		var (
+			count   int
+			exclude []string
+		)
 
-		if len(hh) == 0 {
-			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+		for {
+			if count > maxSyncRequests {
+				return fmt.Errorf("exceeded max sync request limit of %d", maxSyncRequests)
```

Comment:
This should be warn level, not error.

Reply:
Instead of hard-coding `maxSyncRequests` let's use `maxSyncRequests := limit / 24 * 2`.

We also need to inspect the response code of each heartbeat in the bulk response. If all heartbeat responses in the bulk response are error codes, or the bulk response itself has an error code, then we should break and stop sending offline heartbeats.

Reply:
> Instead of hard-coding maxSyncRequests let's use maxSyncRequests := limit / 24 * 2.

So this would restrict it to 2 requests. But if the user has more than 48 heartbeats in the queue, we want to sync them all, right?

Reply:
> We also need to inspect the response code of each heartbeat in the bulk response. If all heartbeat responses in the bulk response are error codes, or the bulk response itself has an error code, then we should break and stop sending offline heartbeats.

Alright, to sum it up and please correct me if I am missing something:
- Response global error: push heartbeats back to queue and abort sync
- All results with invalid status: push heartbeats back to queue and abort sync
- 429/500 result: push heartbeat back to queue and continue sync, but do not send this heartbeat again in this cli execution.
- 400: error log, reject heartbeat and continue sync

Reply:
> 3. 429/500 result: push heartbeat back to queue and continue sync, but do not send this heartbeat again in this cli execution.

The error code doesn't matter, as long as it's not 400. On any error result, put the heartbeat at the back of the queue and continue syncing until the total heartbeats sent reaches `--sync-offline-activity limit`. Each heartbeat sent (failure or success) increases the total count approaching `--sync-offline-activity limit`. If we haven't reached `limit` yet, we can attempt to re-send non-400 heartbeats again in the same cli execution.

Reply:
`--sync-offline-activity limit` is doing the same as your `As a safety measure, to prevent an endless loop, a maximum sync request limit of 100 was added`, so we don't really need the 100 limit.